### PR TITLE
Add PII detection for user and resource IDs

### LINKED ISSUE #1702: Emit a log when a written tuple contains PII
### Checklist

- [X] I agree to the terms within the [OpenFGA Code of Conduct](https://github.com/openfga/.github/blob/main/CODE_OF_CONDUCT.md).

### Describe the problem you'd like to have solved

OpenFGA discourages developers to write PII, but it can't forbid them to do it. 

### Describe the ideal solution

As a way to identify when PII is written, OpenFGA could emit a log entry whenever PII is identified in the user-id or resource-id of a tuple. 

Examples of PII that could be used are emails, usernames, personal identifiers.

### Alternatives and current workarounds

-

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Siddhant-K-code**
## Description

This PR adds PII (Personally Identifiable Information) detection for user and resource IDs. It includes a regular expression to check for email addresses, phone numbers, and social security numbers. If any PII is detected, an error is returned. 

It helps to ensure that sensitive information is not inadvertently stored in the database.

## References

fixes #1702

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: pkg/storage/sqlcommon/sqlcommon.go [hunk 1]
```diff
@@ -326,6 +327,23 @@ func NewDBInfo(db *sql.DB, stbl sq.StatementBuilderType, sqlTime interface{}) *D
 	}
 }
 
+var (
+	emailRegex       = regexp.MustCompile(`^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`)
+	phoneNumberRegex = regexp.MustCompile(`^(\+\d{1,2}\s?)?1?\-?\(?\d{3}\)?[\s.-]\d{3}[\s.-]\d{4}$`)
```
Comment (miparnisari) [inline pkg/storage/sqlcommon/sqlcommon.go:332]: If you copied these regexes from somewhere do you mind documenting a link with the source?

File: pkg/storage/sqlcommon/sqlcommon.go [hunk 2]
```diff
@@ -326,6 +327,23 @@ func NewDBInfo(db *sql.DB, stbl sq.StatementBuilderType, sqlTime interface{}) *D
...
+	ssnRegex         = regexp.MustCompile(`^\d{3}-\d{2}-\d{4}$`)
```
Comment (miparnisari) [inline pkg/storage/sqlcommon/sqlcommon.go:333]: Personal opinion, this looks too premature of a PII check to me, as it's specific to the USA. Are we going to add PII checks for other types of IDs too?

**Review by miparnisari — CHANGES_REQUESTED** [review]

File: pkg/storage/sqlcommon/sqlcommon.go (hunk 2 above)
Comment (Siddhant-K-code) [reply pkg/storage/sqlcommon/sqlcommon.go:333]: I am also struggling, with what to verify under PII, what could be the other patterns we can compare and check against?

**Comment by Siddhant-K-code** [comment]
Putting it in a draft, until we find the answer for https://github.com/openfga/openfga/pull/1706#discussion_r1653872236

**Comment by miparnisari** [comment]
@Siddhant-K-code i think that having the check for email, to start with, is fine.

**Comment by Leonardo-Ferreira** [comment]
my two cents:

1. this should be OPT-IN... those regex checks can really add-up
2. not all emails are PII
3. PII detection is really really hard and elusive... IP alone is not PII, but IP+timestamp CAN BE because technically one IP can only be held by 1 entity at a given point in time

**Pull request closed without merging**